addons/website_livechat/models/mail_channel.py

```diff
@@ -50,8 +50,15 @@ def _get_visitor_history(self, visitor):
         :param visitor: website.visitor of the channel
         :return: arrow separated string containing navigation history information
         """
-        recent_history = visitor.visitor_page_ids[:3] if len(visitor.visitor_page_ids) >= 3 else visitor.visitor_page_ids
-        return ' → '.join(page.page_id.name + ' (' + page.visit_datetime.strftime('%H:%M') + ')' for page in recent_history)
+        history = []
+        count = 0
+        for page in visitor.website_track_ids:
+            if page.page_id:
+                history += page.page_id.name + ' (' + page.visit_datetime.strftime('%H:%M') + ')'
+                count += 1
+            if count == 3:
```

Comment:
if len(history) ==3

Reply:
visitor.website_track_ids.filter('page_id')[:3]

Reply:
what if there is 1000+ tracks, filter will go through all of them. Isn't it an issue ?